### ISSUE TITLE
Target sensor wrong time window offered under some circumstances

### ISSUE THREAD

**Issue by ragg987** [BottlecapDave/HomeAssistant-OctopusEnergy#1017]
### Describe the bug

When I update a target rate sensor through Actions where the requested hours exceeds the cheap rate window (in my example it is the 6hr window for IOG), and I use a weighting the result is non-optimal.

### Reproduction steps

here is the yaml Action and output to reproduce the issue.

YAML
```
action: octopus_energy.update_target_config
target:
  entity_id: binary_sensor.octopus_energy_target_myclimate_intelligent_heating
data:
  target_hours: "7.5"
  target_start_time: "21:00"
  target_end_time: "20:00"
  target_weighting: 2,*
  target_offset: "-00:01:00"
```

RESULT (note the weighting 2 factor is at the expensive rate at 22:00 hours)
```
name: myclimate_intelligent_heating
hours: 7.5
hours_mode: exact_hours
type: Continuous
mpan: xxxx
rolling_target: false
last_rates: true
target_invert_target_rates: false
kind: target_rate
account_id: A-xxxx
offset: -00:01:00
start_time: 21:00
end_time: 20:00
weighting: 2,*
is_target_export: false
icon: mdi:camera-timer
friendly_name: Octopus Energy Target myclimate_intelligent_heating
rates_incomplete: false
target_times: 
- value_inc_vat: 0.267159
  start: '2024-10-10T22:00:00+01:00'
  end: '2024-10-10T22:30:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.267159
  start: '2024-10-10T22:30:00+01:00'
  end: '2024-10-10T23:00:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.267159
  start: '2024-10-10T23:00:00+01:00'
  end: '2024-10-10T23:30:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-10T23:30:00+01:00'
  end: '2024-10-11T00:00:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-11T00:00:00+01:00'
  end: '2024-10-11T00:30:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-11T00:30:00+01:00'
  end: '2024-10-11T01:00:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-11T01:00:00+01:00'
  end: '2024-10-11T01:30:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-11T01:30:00+01:00'
  end: '2024-10-11T02:00:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-11T02:00:00+01:00'
  end: '2024-10-11T02:30:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-11T02:30:00+01:00'
  end: '2024-10-11T03:00:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-11T03:00:00+01:00'
  end: '2024-10-11T03:30:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-11T03:30:00+01:00'
  end: '2024-10-11T04:00:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-11T04:00:00+01:00'
  end: '2024-10-11T04:30:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-11T04:30:00+01:00'
  end: '2024-10-11T05:00:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
- value_inc_vat: 0.070003
  start: '2024-10-11T05:00:00+01:00'
  end: '2024-10-11T05:30:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false

target_times_last_evaluated: 2024-10-10T21:58:51.866913+01:00
overall_average_cost: 0.10943
overall_min_cost: 0.070003
overall_max_cost: 0.267159
current_duration_in_hours: 7.5
current_average_cost: 0.10943420000000002
current_min_cost: 0.070003
current_max_cost: 0.267159
next_time: null
next_duration_in_hours: 0
next_average_cost: null
next_min_cost: null
next_max_cost: null
```

### Expected behaviour

I would expect the target to start at 23:30 and finish 7.5hrs later. This would respect the 2 weighting factor for the first hour.

### Tariff Code

E-1R-INTELLI-VAR-22-10-14-A

### Integration Version

12.2.1

### Home Assistant Version

2024.10.1 HAOS core

### Fresh Install?

Not specified

### Home Assistant Logs

I attach a heavily trimmed log limited to the few seconds between performing the action and the sensor updating.
[home-assistant_myenergi_2024-10-10T21-09-54.419Z.log](https://github.com/user-attachments/files/17334266/home-assistant_myenergi_2024-10-10T21-09-54.419Z.log)

### Confirmation

- [X] I confirm that I cannot find my solution within the [documentation](https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy)
- [X] I confirm that I cannot find my solution within the [FAQ](https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy/faq)

**Comment by ragg987** [comment]
Update, it seems to work correctly (sometimes - this works sometimes and is wrong at other times) if the number of hours is a whole number.
```
action: octopus_energy.update_target_config
target:
  entity_id: binary_sensor.octopus_energy_target_myclimate_intelligent_heating
data:
  target_hours: "8"
  target_start_time: "22:30"
  target_end_time: "20:00"
  target_weighting: "2,*"
  target_offset: "-00:01:00"
```
```
...
target_times: 
- value_inc_vat: 0.070003
  start: '2024-10-10T23:30:00+01:00'
  end: '2024-10-11T00:00:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
...
- value_inc_vat: 0.267159
  start: '2024-10-11T07:00:00+01:00'
  end: '2024-10-11T07:30:00+01:00'
  tariff_code: E-1R-INTELLI-VAR-22-10-14-A
  is_capped: false
...
```

**Comment by BottlecapDave** [comment]
Hello and sorry your encountering this issue. I've managed to recreate the issue, which was down to weighting being used in conjunction with `Find last applicable rates`, and a fix will be available in the next release.

**Comment by BottlecapDave** [comment]
The fix for this issue is now available in [v13.0.0](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/releases/tag/v13.0.0).
